atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/CoordinationService.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.coordination;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+import com.palantir.atlasdb.keyvalue.impl.CheckAndSetResult;
+
+/**
+ * A {@link CoordinationService} is used to agree on values being relevant or correct at a given timestamp.
+ * The sequence of values being agreed should evolve in a backwards consistent manner; that is, if I read a value
+ * for some timestamp TS and make a decision at TS, all future values written to the {@link CoordinationService} should
+ * also ensure that no decision that would be inconsistent with the decision made at TS would be made at TS. In general,
+ * this means that one should avoid changes that affect behaviour of values below the validity bound.
+ *
+ * It is the responsibility of the caller to preserve the above property.
+ */
+public interface CoordinationService<T> {
+    /**
+     * Returns a value that the coordination service has agreed is appropriate at the provided timestamp.
+     *
+     * The value returned by this method may change over time. However, values returned by this method would
```

Comment:
May change here means that we may add some new feature that needs to change some behavior, but it would be applied in a way such that any value read here would be consistent with the behavior we were already using?

Reply:
Yep, that's exactly the idea I was going for. Changes are applied in a way that any values for behaviour at timestamps less than the previous validity bound are preserved.

Within the context of _transactions2, the idea is that you have a `RangeMap` of timestamps to transaction schema you should be using. This is initially `(0, infinity) => 1`, but once installed at some validity `V` (where the previous validity was `V'`) this looks like `(0, V'] => 1, (V', infinity) => 2`

Reply:
In practice, though, any client should treat it as though the bound is `(0, V'] => 1, (V', bound) => 2`, right? It would obviously be dangerous to assume any behavior extends to infinity

Reply:
Right, though generally I think we assume `bound` itself is good - so the part between `[bound+1, infinity)` should not be relied on to stay the same.

Reply:
The main reason I use infinity is that if this turns out to be a part of some Atlas internal schema (so we have something like `{"transactions": "(0, inf) => 1", "otherThing": 1}`) and we want to update the `otherThing` key, we don't have to think about logic for transactions directly.